### DISCUSSION TITLE
[form]表单组件是否可以提供一个自定义错误设置功能

### DISCUSSION THREAD

**Discussion by fangjc1986** [element-plus/element-plus#13922] [Ideas]
# 需求来源

1. 比如说添加一个用户的表单，`username` 前端检查字符串格式，但是提交的时候可能存在 `username` 重复的现象；
2. 当 `username` 重复的时候， 后端会返回类似这样的验证错误信息: `{ username : "此用户名已存在" }` ;
3. 我希望将错误信息放在表单 `username` 输入框的下方以提示用户；

# 当前处理方案

1. 通过 `validator` 的异步校验方式进行校验，但是这样就需要后端提供一个额外的接口专门做校验，且在提交表单的时候的校验错误无法传递到表单上；

# 其他框架已有的处理方案

1. `arco-design` 在 `form` 组件上提供了 `setFields` 的额外功能，用于设置某个字段的当前状态，从而可以将服务器返回的错误信息手动添加到 `form` 表单上；

![image](https://github.com/element-plus/element-plus/assets/21995186/be90c383-acfc-4b54-9252-3152aaa88c9f)

![image](https://github.com/element-plus/element-plus/assets/21995186/3f596b6d-f97f-4b60-97f2-05791e4ae8be)

# 结论与需求

1. 建议 `element-plus` 也可以提供类似的 `setFields` 的方法来手动设置表单状态；

# 实现代码示例

```vue

<script setup lang="ts">
import { ref } from "vue";
import axios from "axios";

const formRef = ref();
const form = ref({
  username: "",
});

axios
  .post("/api/user/save", form.value)
  .then((resp: any) => {
    console.log("保存成功");
  })
  .catch((err: object) => {
    // 当出现表单验证错误时，将错误信息设置到表单中
    // err : { username : "用户名已存在" }
    formRef.value.setFields(
      Object.keys(err).reduce((t, k) => {
        t[k] = {
          status: "error",
          message: err[k],
        };
        return t;
      }, {}),
    );
  });
</script>

<template>
  <el-form ref="formRef">
    <el-form-item prop="username" label="用户名">
      <el-input v-model="form.username"></el-input>
    </el-form-item>
  </el-form>
</template>

<style scoped></style>


```

**Comment by yangmiao7** [comment]


**Reply by fangjc1986** [reply]
已经提交 PR， 不知道什么时候能过

https://github.com/element-plus/element-plus/pull/14017

**Comment by gimjin** [comment]
不是有[自定义表单校验](https://element-plus.org/zh-CN/component/form.html#%E8%87%AA%E5%AE%9A%E4%B9%89%E6%A0%A1%E9%AA%8C%E8%A7%84%E5%88%99)吗？demo 里面就是用 setTimeout 模拟了异步请求。

**Reply by fangjc1986** [reply]
1. 比如自定义校验如果成功了，但是最后提交表单之后，后端接口再次校验发现问题，此时如何展示错误；
2. 一般后端会在提交的时候校验字段的合法性，可能根据各种不同的逻辑进行校验，如果可以将提交返回的校验错误在表单上直接展示，后端就可以不用额外添加一个校验接口；

**Reply by gimjin** [reply]
嗯，如果后端返回失败状态，再执行 validate() 方法就会在错误的表单项下面用红色字提醒。 因为你的自定义 validate rules 已经写好了，程序会重新帮你检查你的表单，而不用手动控制。手动控制后端还要告诉你失败的字段标识位，如果运气不好有很多个表单项失败你还得循环手动设置，很麻烦。

祝你好运。

**Reply by fangjc1986** [reply]
我明白你的意思，可是这样后端不是需要额外添加用于验证的 API 接口吗？

**Reply by fangjc1986** [reply]
这些都可以封装在 axios 的 Response 拦截器中，当表单发起请求的时候，就将 formRef（Form表单组件） 对象添加到 axios 的 config 中；

当返回结果 code 状态为 `字段验证失败` 的时候，后端将统一返回格式： {username: "用户名重复", age: "年龄超出范围"};

之后 axios 可以直接 通过 code 码判断是否为字段验证错误，并且可以利用 config 中的  formRef，`formRef.value.setValidateResults(resp.data);` 来完成表单错误设置，无需在业务层做任何操作；’

**Reply by fangjc1986** [reply]
代码类似：

```js
// 在 axios 中添加响应拦截器
axios.interceptors.response.use( response => {
// 判断是否为表单验证错误
if( resp.data.code === 120 ){
// resp.data.err : {username: "用户名重复", age: "年龄超出范围"}
 response.config.formRef.value.setValidateResults(resp.data.err);
}

// 业务层发起请求
axios.post({
  method: 'post',
  url: '......',
  formRef
}).then(....);
```

**Comment by staoran** [comment]
这个功能好，苦等这个功能许久了😄

**Comment by Arick-a** [comment]
大哥 咱俩需求一模一样   现在已经实现了吗？

**Reply by fangjc1986** [reply]
已经转投 arco-design 了

**Comment by putianyi889** [comment]
你好，请问可以提供一个自行实现该功能的代码示例吗？谢谢

**Reply by staoran** [reply]
```
	/**
	 * 表单实例
	 */
	const ruleFormRef = ref<FormInstance>();

	/**
	 * 表单反显服务端返回的字段错误列表
	 */
	const handleErrors = (errors: Record<string, string[]>) => {
		if (ruleFormRef.value) {
			Object.keys(errors).forEach((key) => {
				const lowerCaseKey = key.charAt(0).toLowerCase() + key.slice(1);

				// :ref="el => formItemRefs['confirmNewPassword'] = el"
				// const formItemRefs = ref<Record<string, any>>({});
				// const field = formItemRefs.value[lowerCaseKey] as any;
				// if (field) {
				// 	field.validateState = 'error';
				// 	field.validateMessage = errors[key][0];
				// }

				ElMessage.error(errors[key][0]);
				const field = ruleFormRef.value?.fields.find((field) => field.prop === lowerCaseKey);
				if (field) {
					field.validateState = 'error';
					nextTick(() => {
						const maxRetries = 10; // 最大递归次数
						let retryCount = 0;
						const updateErrorMessage = () => {
							if (field.$el !== undefined) {
								const itemErr = field.$el.querySelector('.el-form-item__error');
								if (itemErr) {
									itemErr.textContent = errors[key][0];
								} else if (retryCount < maxRetries) {
									// 如果找不到元素，递归调用直到找到为止
									retryCount++;
									setTimeout(updateErrorMessage, 20);
								}
							}
						};
						updateErrorMessage();
					});
				}
			});
		}
	};
```

**Comment by tangzhangming** [comment]
我以为是我没找到，结果是真的没有这个功能 蛋疼的要死

**Comment by xzhdev** [comment]
pr 居然没过，心痛

**Comment by xingyixiang** [comment]
#20981
[Example](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtZm9ybVxuICAgIHJlZj1cInJ1bGVGb3JtUmVmXCJcbiAgICBzdHlsZT1cIm1heC13aWR0aDogNjAwcHhcIlxuICAgIDptb2RlbD1cInJ1bGVGb3JtXCJcbiAgICA6cnVsZXM9XCJydWxlc1wiXG4gICAgbGFiZWwtd2lkdGg9XCJhdXRvXCJcbiAgPlxuICAgIDxlbC1mb3JtLWl0ZW0gbGFiZWw9XCJBY3Rpdml0eSBuYW1lXCIgcHJvcD1cIm5hbWVcIiByZXF1aXJlZD5cbiAgICAgIDxlbC1pbnB1dCB2LW1vZGVsPVwicnVsZUZvcm0ubmFtZVwiIC8+XG4gICAgPC9lbC1mb3JtLWl0ZW0+XG4gICAgPGVsLWZvcm0taXRlbSBsYWJlbD1cIkFjdGl2aXR5IHpvbmVcIiBwcm9wPVwicmVnaW9uXCI+XG4gICAgICA8ZWwtc2VsZWN0IHYtbW9kZWw9XCJydWxlRm9ybS5yZWdpb25cIiBwbGFjZWhvbGRlcj1cIkFjdGl2aXR5IHpvbmVcIj5cbiAgICAgICAgPGVsLW9wdGlvbiBsYWJlbD1cIlpvbmUgb25lXCIgdmFsdWU9XCJzaGFuZ2hhaVwiIC8+XG4gICAgICAgIDxlbC1vcHRpb24gbGFiZWw9XCJab25lIHR3b1wiIHZhbHVlPVwiYmVpamluZ1wiIC8+XG4gICAgICA8L2VsLXNlbGVjdD5cbiAgICA8L2VsLWZvcm0taXRlbT5cbiAgICA8ZWwtZm9ybS1pdGVtPlxuICAgICAgPGVsLWJ1dHRvbiB0eXBlPVwicHJpbWFyeVwiIEBjbGljaz1cInN1Ym1pdEZvcm0ocnVsZUZvcm1SZWYpXCI+XG4gICAgICAgIENyZWF0ZVxuICAgICAgPC9lbC1idXR0b24+XG4gICAgICA8ZWwtYnV0dG9uIEBjbGljaz1cInJlc2V0Rm9ybShydWxlRm9ybVJlZilcIj5SZXNldDwvZWwtYnV0dG9uPlxuICAgIDwvZWwtZm9ybS1pdGVtPlxuICA8L2VsLWZvcm0+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IGxhbmc9XCJ0c1wiIHNldHVwPlxuaW1wb3J0IHsgcmVhY3RpdmUsIHJlZiB9IGZyb20gJ3Z1ZSdcbmltcG9ydCB0eXBlIHsgRm9ybUluc3RhbmNlLCBGb3JtUnVsZXMgfSBmcm9tICdlbGVtZW50LXBsdXMnXG5cbmludGVyZmFjZSBSdWxlRm9ybSB7XG4gIG5hbWU6IHN0cmluZ1xuICByZWdpb246IHN0cmluZ1xufVxuXG5jb25zdCBydWxlRm9ybVJlZiA9IHJlZjxGb3JtSW5zdGFuY2U+KClcbmNvbnN0IHJ1bGVGb3JtID0gcmVhY3RpdmU8UnVsZUZvcm0+KHtcbiAgbmFtZTogJ0hlbGxvJyxcbiAgcmVnaW9uOiAnc2hhbmdoYWknLFxufSlcblxuY29uc3QgcnVsZXMgPSByZWFjdGl2ZTxGb3JtUnVsZXM8UnVsZUZvcm0+Pih7XG4gIG5hbWU6IFtcbiAgICB7IHJlcXVpcmVkOiB0cnVlLCBtZXNzYWdlOiAnUGxlYXNlIGlucHV0IEFjdGl2aXR5IG5hbWUnLCB0cmlnZ2VyOiAnYmx1cicgfSxcbiAgICB7IG1pbjogMywgbWF4OiA1LCBtZXNzYWdlOiAnTGVuZ3RoIHNob3VsZCBiZSAzIHRvIDUnLCB0cmlnZ2VyOiAnYmx1cicgfSxcbiAgXSxcbiAgcmVnaW9uOiBbXG4gICAge1xuICAgICAgcmVxdWlyZWQ6IHRydWUsXG4gICAgICBtZXNzYWdlOiAnUGxlYXNlIHNlbGVjdCBBY3Rpdml0eSB6b25lJyxcbiAgICAgIHRyaWdnZXI6ICdjaGFuZ2UnLFxuICAgIH0sXG4gIF0sXG59KVxuXG4vLyBtb2NrXG5jb25zdCBhcGkgPSAoKSA9PiB7XG4gIHJldHVybiBuZXcgUHJvbWlzZTxSZWNvcmQ8c3RyaW5nLCBzdHJpbmc+PigocmVzb2x2ZSkgPT4ge1xuICAgIHNldFRpbWVvdXQoKCkgPT4ge1xuICAgICAgcmVzb2x2ZSh7XG4gICAgICAgIG5hbWU6ICduYW1lIGlzIGVycm9yJyxcbiAgICAgICAgcmVnaW9uOiAncmVnaW9uIGlzIGVycm9yJyxcbiAgICAgIH0pXG4gICAgfSwgMzAwKVxuICB9KVxufVxuXG5jb25zdCBzdWJtaXRGb3JtID0gYXN5bmMgKGZvcm1FbDogRm9ybUluc3RhbmNlIHwgdW5kZWZpbmVkKSA9PiB7XG4gIGlmICghZm9ybUVsKSByZXR1cm5cbiAgY29uc3QgZGF0YSA9IGF3YWl0IGFwaSgpXG4gIE9iamVjdC5lbnRyaWVzKGRhdGEpLmZvckVhY2goKFtrZXksIG1lc3NhZ2VdKSA9PiB7XG4gICAgICBjb25zdCBmaWVsZCA9IGZvcm1FbC5nZXRGaWVsZChrZXkpXG4gICAgICBpZihmaWVsZCkge1xuICAgICAgICBmaWVsZC52YWxpZGF0ZVN0YXRlID0gJ2Vycm9yJ1xuICAgICAgICBmaWVsZC52YWxpZGF0ZU1lc3NhZ2UgPSBtZXNzYWdlXG4gICAgICB9XG4gIH0pXG59XG5cbmNvbnN0IHJlc2V0Rm9ybSA9IChmb3JtRWw6IEZvcm1JbnN0YW5jZSB8IHVuZGVmaW5lZCkgPT4ge1xuICBpZiAoIWZvcm1FbCkgcmV0dXJuXG4gIGZvcm1FbC5yZXNldEZpZWxkcygpXG59XG48L3NjcmlwdD5cbiIsImVsZW1lbnQtcGx1cy5qcyI6ImltcG9ydCBFbGVtZW50UGx1cyBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBnZXRDdXJyZW50SW5zdGFuY2UgfSBmcm9tICd2dWUnXG5cbmxldCBpbnN0YWxsZWQgPSBmYWxzZVxuYXdhaXQgbG9hZFN0eWxlKClcblxuZXhwb3J0IGZ1bmN0aW9uIHNldHVwRWxlbWVudFBsdXMoKSB7XG4gIGlmIChpbnN0YWxsZWQpIHJldHVyblxuICBjb25zdCBpbnN0YW5jZSA9IGdldEN1cnJlbnRJbnN0YW5jZSgpXG4gIGluc3RhbmNlLmFwcENvbnRleHQuYXBwLnVzZShFbGVtZW50UGx1cylcbiAgaW5zdGFsbGVkID0gdHJ1ZVxufVxuXG5leHBvcnQgZnVuY3Rpb24gbG9hZFN0eWxlKCkge1xuICBjb25zdCBzdHlsZXMgPSBbJ2h0dHBzOi8vcHJldmlldy0yMDk4MS1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2Rpc3QvaW5kZXguY3NzJywgJ2h0dHBzOi8vcHJldmlldy0yMDk4MS1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL3RoZW1lLWNoYWxrL2RhcmsvY3NzLXZhcnMuY3NzJ10ubWFwKChzdHlsZSkgPT4ge1xuICAgIHJldHVybiBuZXcgUHJvbWlzZSgocmVzb2x2ZSwgcmVqZWN0KSA9PiB7XG4gICAgICBjb25zdCBsaW5rID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnbGluaycpXG4gICAgICBsaW5rLnJlbCA9ICdzdHlsZXNoZWV0J1xuICAgICAgbGluay5ocmVmID0gc3R5bGVcbiAgICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignbG9hZCcsIHJlc29sdmUpXG4gICAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2Vycm9yJywgcmVqZWN0KVxuICAgICAgZG9jdW1lbnQuYm9keS5hcHBlbmQobGluaylcbiAgICB9KVxuICB9KVxuICByZXR1cm4gUHJvbWlzZS5hbGxTZXR0bGVkKHN0eWxlcylcbn1cbiIsInRzY29uZmlnLmpzb24iOiJ7XG4gIFwiY29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiBcIkVTTmV4dFwiLFxuICAgIFwianN4XCI6IFwicHJlc2VydmVcIixcbiAgICBcIm1vZHVsZVwiOiBcIkVTTmV4dFwiLFxuICAgIFwibW9kdWxlUmVzb2x1dGlvblwiOiBcIkJ1bmRsZXJcIixcbiAgICBcInR5cGVzXCI6IFtcImVsZW1lbnQtcGx1cy9nbG9iYWwuZC50c1wiXSxcbiAgICBcImFsbG93SW1wb3J0aW5nVHNFeHRlbnNpb25zXCI6IHRydWUsXG4gICAgXCJhbGxvd0pzXCI6IHRydWUsXG4gICAgXCJjaGVja0pzXCI6IHRydWVcbiAgfSxcbiAgXCJ2dWVDb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IDMuM1xuICB9XG59XG4iLCJQbGF5Z3JvdW5kTWFpbi52dWUiOiI8c2NyaXB0IHNldHVwPlxuaW1wb3J0IEFwcCBmcm9tICcuL0FwcC52dWUnXG5pbXBvcnQgeyBzZXR1cEVsZW1lbnRQbHVzIH0gZnJvbSAnLi9lbGVtZW50LXBsdXMuanMnXG5zZXR1cEVsZW1lbnRQbHVzKClcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxBcHAgLz5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvcnVudGltZS1kb21AbGF0ZXN0L2Rpc3QvcnVudGltZS1kb20uZXNtLWJyb3dzZXIuanNcIixcbiAgICBcIkB2dWUvc2hhcmVkXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9AdnVlL3NoYXJlZEBsYXRlc3QvZGlzdC9zaGFyZWQuZXNtLWJ1bmRsZXIuanNcIixcbiAgICBcImVsZW1lbnQtcGx1c1wiOiBcImh0dHBzOi8vcHJldmlldy0yMDk4MS1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2Rpc3QvaW5kZXguZnVsbC5taW4ubWpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXMvXCI6IFwidW5zdXBwb3J0ZWRcIixcbiAgICBcIkBlbGVtZW50LXBsdXMvaWNvbnMtdnVlXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9AZWxlbWVudC1wbHVzL2ljb25zLXZ1ZUAyL2Rpc3QvaW5kZXgubWluLmpzXCJcbiAgfSxcbiAgXCJzY29wZXNcIjoge31cbn0iLCJfbyI6eyJzaG93SGlkZGVuIjp0cnVlLCJzdHlsZVNvdXJjZSI6Imh0dHBzOi8vcHJldmlldy0yMDk4MS1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2Rpc3QvaW5kZXguY3NzIn19)
